crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -412,23 +373,27 @@ where
             })
             .collect::<HashMap<_, _>>();
 
-        caches.iter_mut().for_each(|state| {
-            // Filter-out piece indices that are stored, but should not be as well as clean
-            // `inserted_piece_indices` from already stored piece indices, leaving just those that are
-            // still missing in cache
-            state
-                .stored_pieces
-                .extract_if(|key, _offset| piece_indices_to_store.remove(key).is_none())
-                .for_each(|(_piece_index, offset)| {
-                    state.free_offsets.push_front(offset);
-                });
-        });
+        let mut piece_caches_capacity_used = vec![0i64; caches.backends.len()];
+        // Filter-out piece indices that are stored, but should not be as well as clean
+        // `inserted_piece_indices` from already stored piece indices, leaving just those that are
+        // still missing in cache
+        caches
+            .stored_pieces
+            .extract_if(|key, offset| {
+                if let Some(capacity_used) =
+                    piece_caches_capacity_used.get_mut(usize::from(offset.cache_index))
+                {
+                    *capacity_used += 1;
+                }
```

Comment:
I'd write it as this due to pre-allocation above:
```suggestion
                piece_caches_capacity_used[usize::from(offset.cache_index)] += 1;
```

Thought this is not a correct logic, it should only be increased if the key is not a duplicate (which is exactly what `piece_indices_to_store.remove(key).is_none()` is for below.